virtual_rainforest/core/constants.py

```diff
@@ -1,121 +1,15 @@
-
-    LOGGER.info("Constants class %s.%s registered", model_name, class_name)
-
-
-def check_valid_constant_names(
-    config: Config, model_name: str, class_name: str
-) -> None:
-    """Check that the constant names given in the config are valid.
-
-    This checks that the constants are expected for the specific dataclass that they are
-    assigned to, if not an error is raised.
-
-    Args:
-        config: A validated Virtual Rainforest model configuration object.
-        model_name: Name of the model the constants belong to
-        class_name: Name of the specific dataclass the constants belong to
-
-    Raises:
-        ConfigurationError: If unexpected constant names are used
-    """
-
-    # Extract dataclass of interest from registry
-    constants_class = CONSTANTS_REGISTRY[model_name][class_name]
-
-    # Extract a set of provided constant names
-    provided_names = set(config[model_name]["constants"][class_name].keys())
-
-    # Get a set of valid names
-    valid_names = {fld.name for fld in dataclasses.fields(constants_class)}
-
-    # Check for unexpected names
-    unexpected_names = provided_names.difference(valid_names)
-    if unexpected_names:
-        LOGGER.error(
-            "Unknown names supplied for %s: %s"
-            % (class_name, ", ".join(unexpected_names))
-        )
-        LOGGER.info("Valid names are as follows: %s" % (", ".join(valid_names)))
-        raise ConfigurationError()
-
-    return
-
-
-def load_constants(config: Config, model_name: str, class_name: str) -> Any:
-    """Load the specified constants class.
-
-    Any constants that are supplied for this class in the config are used to populate
-    the class, for all other constants default values are used.
-
-    Args:
-        config: A validated Virtual Rainforest model configuration object.
-        model_name: Name of the model the constants belong to
-        class_name: Name of the specific dataclass the constants belong to
+from dataclasses import dataclass
 
-    Returns:
-        A constants class populated using the information provided in the config
-    """
+from virtual_rainforest.core.constants_class import ConstantsDataclass
 
-    # Extract dataclass of interest from registry
-    constants_class = CONSTANTS_REGISTRY[model_name][class_name]
 
-    # Check if any constants have been supplied
-    if model_name in config and "constants" in config[model_name]:
-        # Checks that constants in config are as expected
-        check_valid_constant_names(config, model_name, class_name)
-        # If an error isn't raised then generate the dataclass
-        return constants_class(**config[model_name]["constants"][class_name])
+@dataclass(frozen=True)
+class CoreConsts(ConstantsDataclass):
+    """Core constants for use across the :mod:`virtual_rainforest` modules."""
 
-    # If no constants are supplied then the defaults should be used
-    return constants_class()
+    k_to_c: float = 273.15
```

Comment:
Agreed ^^

Reply:
I see what you mean but do we really want multiple constants frameworks? And really, the whole simulation is one big red button 😄 . I'm not convinced we should bend over backwards to stop someone who is running a massive ecological simulation from redefining gravity or whatever (and secretly, I kinda want to see what happens).

Having said that - maybe we can update this to allow _some_ constants to be class attributes, not instance attributes, which I think will make them unconfigurable.

Reply:
I mean I'm fine with changing things like the strength of gravity (though that would be tricky as things like flow rates will assume gravity's strength implicitly), I just feel like allowing people to change the number of millimetres in a meter would be bizarre

Reply:
If we ever did a "what is the most sensitive parameter" style sensitivity analysis sweep we would want some way to exclude universal constants + unitary conversions

Reply:
But when would a user ever want to redefine the Celsius-to-Kelvin conversion factor? That is just odd. You don't need a framework for these constants -- just define them somewhere.

For physical units, I think you should also try to use `scipy.constants` wherever possible.

Reply:
The main motivation here is that constants are organised so that they come through one route, rather than having to remember which bits come from where. I agree that `scipy.constants` is great, but it is predominantly physics and quite limited for constants we need for the VR.

We can explicitly set constants as being non-configurable using class variables though. Does this seem sensible:

```python
from dataclasses import dataclass
from typing import ClassVar

from scipy.constants import zero_Celsius

from virtual_rainforest.core.constants_class import ConstantsDataclass

@dataclass(frozen=True)
class Test(ConstantsDataclass):

    zero_Celsius: ClassVar[float] = zero_Celsius
    b: int = 3
```

And hence:

```python
In [288]: Test(zero_Celsius=263.15)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[288], line 1
----> 1 Test(zero_Celsius=263.15)

TypeError: Test.__init__() got an unexpected keyword argument 'zero_Celsius'
```

We could update `ConstantsDataclass.from_config` to report attempts to configure things that should never change.

Reply:
@davidorme definitely agree that that is a good solution in general. However, probably should mention that `zero_Celsius` is no longer going to be used in the model as I've switched to using `scipy.constants.convert_temperature()` to switch between temperature scales rather than writing out my temperature conversions myself

Reply:
That's fine - it can sit there as a demo for now. We're going to have to review all of this.